src/wrappers/themis/wasm/test/test.js

```diff
@@ -77,4 +77,161 @@ describe('wasm-themis', function() {
+
+            let wrappedMessageB = server.wrap(messageB)
+            let unwrappedMessageB = client.unwrap(wrappedMessageB)
+            assert.deepStrictEqual(unwrappedMessageB, messageB)
+
+            client.destroy()
+            server.destroy()
+        })
+        it('handles unknown clients', function() {
+            let keys = makeKeyMaterial()
+            let client = new themis.SecureSession(randomID, keys.client.privateKey, keys.clientCallback)
+            let server = new themis.SecureSession(serverID, keys.server.privateKey, keys.serverCallback)
+
+            let request = client.connectionRequest()
+            assert.throws(() => server.negotiateReply(request),
+                expectError(ThemisErrorCode.SSESSION_GET_PUB_FOR_ID_CALLBACK_ERROR)
+            )
+
+            client.destroy()
+            server.destroy()
+        })
+        it('handles unknown servers', function() {
+            let keys = makeKeyMaterial()
+            let client = new themis.SecureSession(clientID, keys.client.privateKey, keys.clientCallback)
+            let server = new themis.SecureSession(randomID, keys.server.privateKey, keys.serverCallback)
+
+            let request = client.connectionRequest()
+            let reply = server.negotiateReply(request)
+            assert.throws(() => client.negotiateReply(reply),
+                expectError(ThemisErrorCode.SSESSION_GET_PUB_FOR_ID_CALLBACK_ERROR)
+            )
+
+            client.destroy()
+            server.destroy()
+        })
+        it('does not allow empty client ID', function() {
+            let keyPair = new themis.KeyPair()
+            assert.throws(() => new themis.SecureSession(emptyArray, keyPair.privateKey, function(){}),
+                expectError(ThemisErrorCode.INVALID_PARAMETER)
+            )
+        })
+        it('does not allow empty message', function() {
+            let keys = makeKeyMaterial()
+            let client = new themis.SecureSession(clientID, keys.client.privateKey, keys.clientCallback)
+            let server = new themis.SecureSession(serverID, keys.server.privateKey, keys.serverCallback)
+
+            let data = client.connectionRequest()
+            let peer = server
+            while (!(client.established() && server.established())) {
+                data = peer.negotiateReply(data)
+                if (peer == server) {
+                    peer = client
+                } else {
+                    peer = server
+                }
+            }
+
+            assert.throws(() => server.wrap(emptyArray),
+                expectError(ThemisErrorCode.INVALID_PARAMETER)
+            )
+
+            client.destroy()
+            server.destroy()
+        })
+        it('does not allow public keys in constructor', function() {
+            let keyPair = new themis.KeyPair()
+            assert.throws(() => new themis.SecureSession(clientID, keyPair.publicKey, function(){}),
```

Comment:
add test with empty `clientID` or empty key

Reply:
There's a test for an empty client ID:

https://github.com/cossacklabs/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L169-L174

The users are also not allowed to pass empty keys. We expect only `PrivateKey` instances and they cannot be constructed from an empty array:

https://github.com/cossacklabs/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L59-L62

Reply:
can we test that private keys cannot be constructed from empty strings?

Reply:
> can we test that private keys cannot be constructed from empty strings?

We do:

https://github.com/ilammy/themis/blob/5c01c86dd5b674a9b790c5ebd182f97904b53675/src/wrappers/themis/wasm/test/test.js#L69-L72

They cannot be constructed from _any_ string, not just empty ones.

I can add a test for empty string specifically, if you wish.